### PR TITLE
Update GS backend to comply with new io integration test suite. fixes…

### PR DESCRIPTION
**Pull request by funkyshu**
… #172

### CODE REVIEW

File: backend/gs/file.go [hunk 1]
```diff
@@ -25,84 +25,304 @@ const (
...
+		// close reader
+		if err := f.reader.Close(); err != nil {
+			return utils.WrapCloseError(err)
 		}
+	}
 
-		f.tempFile = nil
+	// finalize writer
+	if f.gcsWriter != nil {
+		// close s3Writer
```
Comment (seeroush) [inline backend/gs/file.go:72]:
```suggestion
		// close gcsWriter
```

File: backend/gs/file.go [hunk 2]
```diff
@@ -25,84 +25,304 @@ const (
...
+		if err := f.gcsWriter.Close(); err != nil {
+			return utils.WrapCloseError(err)
+		}
+	} else if f.tempFileWriter != nil { // s3Writer is nil but tempFileWriter is not nil (seek after write, write after seek)
```
Comment (seeroush) [inline backend/gs/file.go:76]:
```suggestion
	} else if f.tempFileWriter != nil { // gcsWriter is nil but tempFileWriter is not nil (seek after write, write after seek)
```

File: backend/gs/file.go [hunk 3]
```diff
@@ -25,84 +25,304 @@ const (
...
+		// write tempFileWriter to s3
```
Comment (seeroush) [inline backend/gs/file.go:77]:
```suggestion
		// write tempFileWriter to gcs
```

File: backend/gs/file.go [hunk 4]
```diff
@@ -25,84 +25,304 @@ const (
...
+	if f.reader != nil {
+		err := f.reader.Close()
+		if err != nil {
+			return 0, utils.WrapSeekError(err)
+		}
+
+		f.reader = nil
+	}
+
+	// invalidate s3Writer
```
Comment (seeroush) [inline backend/gs/file.go:215]:
```suggestion
	// invalidate gcsWriter
```

File: backend/gs/file.go [hunk 5]
```diff
@@ -25,84 +25,304 @@ const (
...
+	if f.gcsWriter != nil {
+		// cancel s3Writer
```
Comment (seeroush) [inline backend/gs/file.go:217]:
```suggestion
		// cancel gcsWriter
```

File: backend/gs/file.go [hunk 6]
```diff
@@ -25,84 +25,304 @@ const (
...
+	// write to gcs
+	if f.gcsWriter != nil {
+		// write to s3
+		gcsWritten, err := f.gcsWriter.Write(data)
+		if err != nil {
+			return 0, utils.WrapWriteError(err)
+		}
 
+		// ensure both writes are the same
+		if written != gcsWritten {
```
Comment (seeroush) [inline backend/gs/file.go:267]: I might be missing the reasoning behind the dual-writes. Mind giving a brief description on why that's necessary?

File: backend/gs/file.go [hunk 7]
```diff
@@ -25,84 +25,304 @@ const (
...
+			}
+
+			// seek to cursorPos
+			if _, err := f.tempFileWriter.Seek(f.cursorPos, 0); err != nil {
+				return err
+			}
+		}
+	}
+
+	// if we haven't seeked yet, we need to get the s3Writer
```
Comment (seeroush) [inline backend/gs/file.go:301]:
```suggestion
	// if we haven't seeked yet, we need to get the gcsWriter
```

File: backend/gs/file.go [hunk 8]
```diff
@@ -25,84 +25,304 @@ const (
...
+			ctx, cancel := context.WithCancel(f.fileSystem.ctx)
+			f.cancelFunc = cancel
+
+			// get object handle
+			handle, err := f.getObjectHandle()
+			if err != nil {
+				return err
+			}
+
+			// get s3Writer
```
Comment (seeroush) [inline backend/gs/file.go:314]:
```suggestion
			// get gcsWriter
```

**Review by seeroush — CHANGES_REQUESTED** [review]

File: backend/gs/file.go (hunk 1 above)
Comment (funkyshu) [reply backend/gs/file.go]: Fixed. Obviously this was largely a cut/paste from the s3 implemenation.

File: backend/gs/file.go (hunk 6 above)
Comment (funkyshu) [reply backend/gs/file.go:276]: Updated comments in https://github.com/C2FO/vfs/pull/173/commits/ff76f1ccde3c5b5321677b20d6bb83373001894a

**Review by dustin-schnee — APPROVED** [review]

**Review by seeroush — APPROVED** [review]

**Pull request merged**